### PR TITLE
feat: Mutual TLS / Client Certificate Support for WebDAV Sync

### PR DESCRIPTION
**Pull request by cbueth**
Hey, thanks for the app, really appreciate it. I recently moved from iOS back to Android and changed my expenses app on the go. This PR adds mTLS to the WebDav syncronisation.

Tldr;

- Add client certificate selection via Android KeyChain with certificate alias stored per account
- Cache SSLSocketFactory for OkHttp connection reuse (like in davx5 PR #1942)
- Use HTTPS/1.1, not HTTP/2 for mTLS, add timeout recovery with HEAD verification
- Auto-detect and fallback to Class 1 locking when server doesn't support locks
- Update donut library from custom fork to official app.futured.donut release

**Motivation**

I want to use this app on multiple devices and users, but my WebDAV server is protected with mutual TLS (mTLS) requiring client certificates.

**Testing**

Tested on a physical test device (Pixel 3a in tethered mode). Full sync cycle tested successfully, without encryption. I am happy to receive feedback on edge cases.

I have not manually tested the case when a certificate is removed from the device after configuration, the app may need user intervention to reconfigure. This is an area for feedback. The certificate selector UI could also be moved to an "Advanced" section if it should not be normally shown - see screenshot down.

I have added tests files for the `CertificateHelper` and the `X500PrincipalHelper`. They can be moved if there's a more appropriate location.

**Implementation Details**

The implementation adds a certificate selector to the WebDAV setup dialog where users can select a client certificate from the Android KeyChain (much simpler than managing it ourselfes). The certificate alias is stored per account and passed through to the SSL configuration. The user can also remove the certificate selection at (re)configuration.
When implementing in my first version, I got timeouts when using PUT, even though the server had received the file in no time. Digging a bit I noticed each request was creating a new `SSLContext` and `SSLSocketFactory`, which prevented `OkHttp` from reusing connections. With mTLS this meant a full certificate handshake on every request, really slowing down sync. THis is why I preferred to cache `SSLSocketFactory` instances keyed by certificate alias, allowing `OkHttp`'s connection pool to work correctly. I stumbled across the approach implemented in davx5-ose PR #1942 (https://github.com/bitfireAT/davx5-ose/pull/1942) which caches `SSLSocketFactories` to allow `okhttp` HTTPS connection reuse. The HTTP/2 protocol does not properly support client certificates, so the implementation forces HTTP/1.1 when a client certificate is configured to ensure reliable operation.
- As a fallback, after a PUT timeout, the implementation now verifies the file exists on the server using a HEAD request. If the file exists, the upload is considered successful.
- Some WebDAV servers with mTLS enabled might not support Class 2 locking. The implementation auto-detects this and falls back to Class 1 (lock file-based) locking when the server doesn't respond to LOCK requests. The implementation clears lock state when an account is reconfigured and handles timeout gracefully during unlock in Class 1 mode.
- The `SSLSocketFactory` caching uses `SoftReference` to allow garbage collection under memory pressure.

**Dependencies**
This PR updates the `donut` library from the custom fork to the official `app.futured.donut` release (version 2.2.4) to resolve build issues with the fork. The compose UI in `tidBits.kt` and `CriterionReachedDialogFragment.kt` is updated to remove the `strokeCap` parameter which changed in the newer version. I think the build issues came from version differences. So feel free to keep the fork version if there is a good reason to keep it.
I have also added JUnit and Truth testing libraries  to webdav/build.gradle for unit test support.
For logging I used `timber.log.Timber` but can also replace it with a preferred alternative.

**Screenshot**

<details><summary>screenshot of selector with client certificate selector</summary>

Screenshot_20260406-160217_My Expenses Debug

</details>

I'd be happy to get some feedback on the PR, if you are interested to have this feature and if there is still something to work on, or if I should reopen on another branch. It is a working version, but I have not run any integration tests, just the manual test with two budget books, no encryption and few manual entries.

Best,
Carlson

### CODE REVIEW

**Comment by mtotschnig** [comment]
@cbueth Thanks a lot, this looks very promising. Could you remove the changes related to the donut library and merge master. I have updated to the latest upstream donut that does support the strokeCap parameter. Do you have a suggestion on how I could test this? Can you grant test access on your server, or explain how one could be set up locally on Linux?

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt [hunk 1]
```diff
@@ -126,7 +130,9 @@ class WebDavBackendProvider @SuppressLint("MissingPermission") internal construc
             super.lock()
         } else {
             if (!webDavClient.lock(accountUuid)) {
-                throw IOException("Backend cannot be locked")
+                fallbackToClass1 = true
+                accountManagerRef.setUserData(accountRef, KEY_WEB_DAV_FALLBACK_TO_CLASS1, "1")
```
[the remark this answers is not in the document]
Comment (mtotschnig) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: I was wondering along similar lines. During WebDav Setup, we already autodetect Class 2 support in method testClass2Locking, so I am not sure if we need this new logic. What problems did you encounter, that this should solve?

[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: For some reason this detection did not work for my setup. It tried to lock and timed out. Manually checking this, my test server had it deactivated and the app still used the lock method.
This was before I defaulted to HTTP/1.1, on HTTP/2. Maybe it does work on HTTP/1.1. I could remove this and test. Do you have any hint regarding this?

[the remark this answers is not in the document]
Comment (mtotschnig) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: I do not have current regular experience with WebDAV servers, unfortunately. When I implemented the Class 2 check, I used Apache's WebDAV module, which at that time, was the only WebDAV server with Class2 lock support, that was available to me. If I had access to other servers now, I would test again.

**Comment by cbueth** [comment]
Hey @mtotschnig, thanks for the welcoming response. I will drop the donut changes (2996c1e273616c73d5d2f154b8cac16009bef641) and verify this builds on my end.

For testing, having a server setup with mTLS available is indispensable. The user then needs the client certificate in the Android KeyChain one selects at setup. In my setup the WebDAV is provided by a Nextcloud instance, but that can be simplified for testing. The setup on NGINX is explained in detail here: https://docs.nginx.com/nginx-instance-manager/system-configuration/secure-traffic/#mutual-client-certificate-authentication-setup-mtls. Other common web servers also support this. Without knowing your testing setup, proxying your testing server might be the simplest approach.
When I find a moment I will test the removed manual fallback and reply to the other conversations. Other than that, have a nice weekend for now.

**Comment by cbueth** [comment]
For now, I rebased. My local build works without any change. Thank you for updating `donut` to latest.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: Thanks for letting me know.
Would it be desirable to have any testing for WebDAV, with or without mTLS? What is the current testing setup?

File: webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt [hunk 2]
```diff
@@ -62,6 +88,16 @@ class WebdavSetup : EditActivity() {
             Utils.getTextWithAppName(
                 this, R.string.description_webdav_url
             )
+        binding.btnSelectClientCert.setOnClickListener {
+            KeyChain.choosePrivateKeyAlias(this, { alias ->
```
Comment (mtotschnig) [inline webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt:100]: If this returns null, because there is no key available, there should some kind of feedback to the user, like a snackbar informing the user. Otherwise, as happened to me, they are kept wondering if the button is broken.

File: webdav/src/main/res/layout/setup_webdav.xml [hunk 3]
```diff
@@ -80,6 +80,49 @@
                     android:inputType="textPassword"
                     android:maxLines="1" />
             </com.google.android.material.textfield.TextInputLayout>
+            <!-- Client Certificate (mTLS) -->
```
Comment (mtotschnig) [inline webdav/src/main/res/layout/setup_webdav.xml:91]: I would suggest to hide this section under an Advanced toggle, similar to how on the onboarding screen, some options are not shown and are revealed via the More options button.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtotschnig) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: I was able to set up an environment to test your PR with https://github.com/hacdias/webdav and Nginx.

File: webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt [hunk 4]
```diff
@@ -102,4 +115,72 @@ object CertificateHelper {
...
+    fun createSslSocketFactory(
+        context: Context,
+        clientCertAlias: String?,
+        trustedCertificate: X509Certificate?
+    ): Pair<SSLSocketFactory, X509TrustManager> {
+        val cached = getCachedSocketFactory(clientCertAlias)
+        if (cached != null) {
+            Timber.d("Using cached SSLSocketFactory for clientCertAlias=%s", clientCertAlias)
+            return cached.sslSocketFactory to cached.trustManager
+        }
```
[the remark this answers is not in the document]
Comment (mtotschnig) [reply webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt]: This is exactly the problem I run into during testing. I connect to my local Nginx which has a self-signed server certificate, after selecting the client certificate from Android. The first attempt fails, UI shows the info about untrusted certificate and lets me check that I want to trust. Then the next login attempt still fails, because the cached socket factory is used, instead of creating a new one, which uses both the clientCertAlias and the trustedCertificate

File: myExpenses/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -492,6 +492,14 @@
     <string name="description_webdav_url">Please enter the URL to a special folder on your WebDAV server. {app_name} will store its data inside this folder, it won\'t touch anything outside of it.</string>
     <string name="description_webdav_trust_certificate">The website uses a certificate, which is not trusted by Android.</string>
     <string name="label_trust_certificate">Trust this certificate</string>
+    <string name="label_client_certificate">Client certificate (mTLS)</string>
```
Comment (mtotschnig) [inline myExpenses/src/main/res/values/strings.xml]: You also added this to the webDav module, which is a smart thing to do. So I would remove them from the main module, as well as all other resources that you already copied to webdav.

File: webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt]: I changed the cache key to a composite of both `clientCertAlias` and `trustedCertificate` (via `Integer.toHexString(cert.hashCode())`). This ensures that when a user trusts a new certificate, a new socket factory is created that includes both the client cert and the newly trusted server cert. See 96737159f1.

File: webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt [hunk 6]
```diff
@@ -102,4 +115,72 @@ object CertificateHelper {
...
+
+    @JvmStatic
+    @Throws(ClientCertMissingException::class)
+    fun createKeyManager(context: Context, alias: String): KeyManager {
+        val privateKey = KeyChain.getPrivateKey(context, alias)
+            ?: throw ClientCertMissingException()
+
+        val certificateChain = KeyChain.getCertificateChain(context, alias)
+            ?: throw ClientCertMissingException()
+
```
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/client/CertificateHelper.kt:141]: Now both calls are wrapped into try-catch blocks, catching `KeyChainException` and `InterruptedException`, and re-throwing as `ClientCertMissingException(cause)` with the original exception as cause. Also updated `ClientCertMissingException` to accept an optional `cause` parameter. See 9673715.

File: myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/SyncViewModel.kt [hunk 7]
```diff
@@ -138,6 +138,18 @@ open class SyncViewModel(application: Application) : ContentResolvingAndroidView
                 KEY_SYNC_PROVIDER_USERNAME,
                 userData.getString(KEY_SYNC_PROVIDER_USERNAME)
             )
+            userData.keySet().forEach { key ->
+                if (key != KEY_SYNC_PROVIDER_URL && key != KEY_SYNC_PROVIDER_USERNAME) {
+                    accountManager.setUserData(account, key, userData.getString(key))
+                }
+            }
```
[the remark this answers is not in the document]
Comment (cbueth) [reply myExpenses/src/main/java/org/totschnig/myexpenses/viewmodel/SyncViewModel.kt:152]: After iterating over keys present in the bundle, it now also iterates over a set of known WebDAV keys (`KEY_WEB_DAV_CERTIFICATE`, `KEY_WEB_DAV_FALLBACK_TO_CLASS1`, `KEY_ALLOW_UNVERIFIED`, `KEY_CLIENT_CERT_ALIAS`) and explicitly calls `accountManager.setUserData(key, null)` for any that are absent from the bundle, which removes them from the account. See 10c92de.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: Regarding this issue: I removed `accountManagerRef.setUserData(accountRef, KEY_WEB_DAV_FALLBACK_TO_CLASS1, "1")` from the `lock()` method. The `fallbackToClass1` field is now in-memory only for the duration of the `WebDavBackendProvider` instance (a single sync session). This avoids permanently downgrading a server to Class 1 due to transient failures. The Class 2 capability is still autodetected during WebDAV setup via `testClass2Locking`, and the result is persisted through the setup flow's existing `NotCompliantWebDavException(isFallbackToClass1=true)` pathway, see 9673715.

File: webdav/src/main/java/org/totschnig/webdav/sync/client/WebDavClient.kt [hunk 8]
```diff
@@ -116,7 +127,18 @@ class WebDavClient(
...
+            resource.put(requestBody!!, if (withLock) buildIfHeader(parent.location) else null)
+        } catch (e: IOException) {
+            if (e.message?.contains("timeout", ignoreCase = true) == true ||
+                e.cause?.message?.contains("timeout", ignoreCase = true) == true) {
+                if (resource.exists()) {
+                    Timber.d("Upload timed out but file exists on server: %s", fileName)
+                    return resource
+                }
+            }
+            throw e
```
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/client/WebDavClient.kt:140]: For timout detection, I replaced all three string-matching timeout checks with a type-based `IOException.isTimeout()` extension function, in 9673715.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt [hunk 9]
```diff
@@ -136,13 +142,31 @@ class WebDavBackendProvider @SuppressLint("MissingPermission") internal construc
...
             throw IOException(e)
+        } catch (e: IOException) {
+            if (e.message?.contains("timeout", ignoreCase = true) == true ||
+                e.cause?.message?.contains("timeout", ignoreCase = true) == true) {
+                if (!lockFile.exists()) {
+                    Timber.d("Lock file deletion timed out but file does not exist, continuing")
+                    return
+                }
+            }
+            throw e
```
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt:154]: See above.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt [hunk 10]
```diff
@@ -136,13 +142,31 @@ class WebDavBackendProvider @SuppressLint("MissingPermission") internal construc
...
         if (fallbackToClass1) {
-            super.unlock()
+            try {
+                super.unlock()
+            } catch (e: IOException) {
+                if (e.message?.contains("timeout", ignoreCase = true) == true ||
+                    e.cause?.message?.contains("timeout", ignoreCase = true) == true) {
+                    Timber.d("Unlock timed out in Class 1 mode, proceeding anyway")
+                    return
+                }
```
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt:167]: See above.

File: webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt [hunk 11]
```diff
@@ -90,8 +126,16 @@ class WebdavSetup : EditActivity() {
...
+                        Toast.makeText(this, R.string.error_client_cert_missing, Toast.LENGTH_LONG).show()
+                        mClientCertAlias = null
+                        updateClientCertUi()
+                    }
                     else -> {
-                        binding.edtUrl.error = Utils.getCause(throwable).message
+                        val cause = Utils.getCause(throwable)
+                        val errorMessage = cause.message ?: cause.javaClass.simpleName ?: "Unknown error"
+                        android.util.Log.e("WebdavSetup", "WebDAV validation failed: $errorMessage", throwable)
+                        binding.edtUrl.error = errorMessage
```
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt:150]: In d60acd5, `android.util.Log.e(...)` was replaced with `Timber.e(throwable, ...)`.

File: webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/sync/WebDavBackendProvider.kt]: > I was wondering along similar lines. During WebDav Setup, we already autodetect Class 2 support in method testClass2Locking, so I am not sure if we need this new logic. What problems did you encounter, that this should solve?

The in-memory fallback handles servers that intermittently fail Class 2 locking during sync (not just at setup time). Without it, a transient lock failure would crash that sync session. With it, the session degrades gracefully to Class 1 and retries Class 2 on the next sync. If you prefer, I can remove the in-memory fallback entirely and let it fail. Let me know.

File: webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt (hunk 2 above)
Comment (cbueth) [reply webdav/src/main/java/org/totschnig/webdav/dialog/WebdavSetup.kt:100]: I added a `Snackbar` with the message "No client certificate available. Please install one in your system credentials." when the alias callback receives `null`, see d60acd5.

File: webdav/src/main/res/layout/setup_webdav.xml (hunk 3 above)
Comment (cbueth) [reply webdav/src/main/res/layout/setup_webdav.xml:91]: Agreed. I added a "More options" button above the client certificate section. The section starts with `visibility="gone"` and is toggled by tapping the button. In reconfigure mode, if a client cert alias is already configured, the section is shown automatically. d60acd527b

File: myExpenses/src/main/res/values/strings.xml (hunk 5 above)
Comment (cbueth) [reply myExpenses/src/main/res/values/strings.xml]: Thanks for the reminder, I moved the remaining strings to `webdav/src/main/res/values/strings.xml`. d60acd527b

**Comment by cbueth** [comment]
Thanks @mtotschnig, for the patience and again for the feedback. I have rebased my feature branch and integrated the feedback. It was rather straight forward.

**Comment by mtotschnig** [comment]
@cbueth I needed 2 fixes in order to make your branch build:
https://github.com/cbueth/MyExpenses/compare/feature/mutual-tls...mtotschnig%3AMyExpenses%3Afeature/mutual-tls
If you pull this commit, I'll approve and merge, after asking Copilot to review again. I can tidy up the resources after that on the master branch.

**Comment by cbueth** [comment]
Et voilà! Thanks

**Comment by cbueth** [comment]
Thank you, Michael. I am looking forward to the upcoming release :)
